Implement map import progress

achou11 opened the pull request:
Closes #1017
Closes #1018

Notes:

- in addition to adding map import progress, this became a massive refactor of the map server integration as a whole (thanks Gregor for leading that work!). 
- also makes various UI improvements for different map management screens

- Potential improvements:
  - think there are still some rough spots with the error handling, and I haven't been able to fully test various error scenarios just yet.

Preview:

- Successful import

  ![map-import-pt-1](https://user-images.githubusercontent.com/18542095/201795731-7a1c5805-7f8c-4ced-b790-22bf14979107.gif)

- Switching maps
  
  ![map-import-pt-2](https://user-images.githubusercontent.com/18542095/201795878-de6fbe40-96be-495b-95eb-cb5de3bc0688.gif)

- General screenshot

  <img width="300" alt="image" src="https://user-images.githubusercontent.com/18542095/201795960-7bb8b9c9-6e00-4b4a-9747-699ab5e3f133.png">

ErikSin committed:
chore: typings and polyfille

achou11 committed:
fix package lock

achou11 committed:
WIP: getting some data back but in the wrong place

achou11 committed:
Get basic progress stuff working

achou11 committed:
Update translations

achou11 committed:
Add progress bar and better error handling

achou11 committed:
update translations

achou11 committed:
small cleanup

achou11 committed:
fix comment

achou11 committed:
fix name of Imports type

achou11 committed:
forgot to save change for last commit

achou11 committed:
create contexts for backgrounded map imports

achou11 committed:
Merge branch 'develop' into new-map-import-progress

achou11 committed:
improvements to state updates and structure

achou11 committed:
major refactoring and refinements related to bottom sheet

achou11 committed:
update translations

ErikSin commented on src/frontend/context/BackgroundedMapImportsContext.tsx:
```diff
@@ -0,0 +1,70 @@
+import * as React from "react";
+
+type BackgroundedMapImportsState = Record<string, string>;
```
For Readability: `type BackgroundedMapImportsState = Record<MapServerStyleInfo["id"], string>;`

ErikSin replied:
I suppose we should actually do `Record<MapServerStyleInfo["id"], Imports["id"]>`

achou11 replied:
need to fix a type in the map server dep in order for `Imports["id"]` to work properly, so probably will leave that part as string for now. see 33024293f9b8a2803b5f16822674f34c818da19b

ErikSin commented on patches/react-native-fetch-api+3.0.0.patch:
```diff
@@ -0,0 +1,15 @@
+diff --git a/node_modules/react-native-fetch-api/src/Fetch.js b/node_modules/react-native-fetch-api/src/Fetch.js
```
Can we document this patch somewhere?

achou11 replied:
Does it help to document as a comment in the patch itself?

ErikSin replied:
I suppose so. Or we could start a readme for all the patches? What do you think?

gmaclennan replied:
There [is a readme for all the patches](https://github.com/digidem/mapeo-mobile/blob/develop/patches/README.md) that I had been keeping up-to-date. Seems like it is a bit behind now, we should document everything there.

ErikSin replied:
@gmaclennan nice! great, so lets do it there!

achou11 replied:
see 2b517c24023df1b814a4c77055f9268d93b6a412

achou11 committed:
Merge branch 'develop' into new-map-import-progress

achou11 committed:
update styling

ErikSin commented on src/frontend/sharedComponents/BGMapCard.tsx:
```diff
@@ -23,86 +29,258 @@ const m = defineMessages({
+  total: number;
+} | null {
+  return (
+    importStatus === null ||
```
We want to show progress bar when null?

achou11 replied:
yeah it's kind of confusing and maybe there's a better way to represent the state using an explicit status string, but basically when the `importStatus` is null, it could mean that we've started an import but haven't received any events back yet. In that case, we want to show the progress bar because it will be in the indeterminate state.

the reason I'm using a null value at all is because if the map's already been imported some time before e.g. yesterday, then there's no need to have an importStatus associated for it, hence null.

would it be helpful to add an import status that represents this "waiting" situation? e.g. extends the `ImportStatus` type so that there's a variant like `{ status: 'waiting' }`?

ErikSin replied:
Yeah, it can't hurt to have that explicit status of `waiting`. And looking through the code, it would simplify some of your null check logic

achou11 replied:
see 4012d1da791f78ebcea8a7a0f74189ac938e6f37

achou11 replied:
did additional work on this via 8bcad51dd905de3f2a7c9e21ba007b9aeb7fe0a7 and think it's much easier to follow now. basically removes the `null` variant and updates the wording of `"waiting"` to `"idle"`

ErikSin commented on src/frontend/screens/Settings/MapSettings/BackgroundMaps.tsx:
```diff
@@ -89,88 +130,121 @@ export const BackgroundMaps: NativeNavigationComponent<"BackgroundMaps"> = ({
-          ))
-        )}
-      </ScrollView>
+  const bottomSheetContentProps: React.ComponentProps<typeof BottomSheetContent> =
```
Too many dependencies to memoize this gracefully I assume?

achou11 replied:
yeah probably not worth memoizing. didn't observe any notable issues as is

achou11 committed:
Merge branch 'develop' into new-map-import-progress

achou11 committed:
fix add button styling

achou11 committed:
undo accidental initial bottom sheet state change

ErikSin reviewed (commented):
Just did an code review, but I will try it on my phone later!

achou11 committed:
better handling of import error event

achou11 committed:
disable backdrop press

achou11 committed:
add waiting variant to import status state

achou11 committed:
improve types using MapServerStyleInfo

achou11 committed:
add description for react-native-fetch-api patch

achou11 committed:
Merge branch 'develop' into new-map-import-progress

achou11 committed:
improvements to state in BgMapCard

achou11 committed:
fix scroll bug in BackgroundMaps

achou11 committed:
add comment explaining why not bottom sheet modal

achou11 committed:
add comment about pressBehavior prop

achou11 committed:
fix issue with children prop in bottom sheet content

achou11 committed:
use BottomSheetModal instead of BottomSheet

achou11 committed:
update bottom sheet dep to latest v2

achou11 committed:
update descriptions for a couple of translations

achou11 committed:
use flat list instead of scrollview

achou11 committed:
improve list styles

achou11 commented:
Anyone who tries this out on a real device, there's currently an issue where if an import error occurs, while the bottom sheet appears, some infinite rendering or something occurs and locks up the UI, eventually crashing. So yeah, error handling still needs work (think it's related to the event source stuff that I'm not handling properly. not the easiest to reproduce either...)

ErikSin commented:
> Anyone who tries this out on a real device, there's currently an issue where if an import error occurs, while the bottom sheet appears, some infinite rendering or something occurs and locks up the UI, eventually crashing. So yeah, error handling still needs work (think it's related to the event source stuff that I'm not handling properly. not the easiest to reproduce either...)

Yeah, I have not been able to reproduce this error. Do you have any ideas of what things I should try to reproduce, so I can help debug?

achou11 commented:
> Yeah, I have not been able to reproduce this error. Do you have any ideas of what things I should try to reproduce, so I can help debug?

using a bad mbtiles file would probably be the most effective approach. we have a few here that we should try using, since we know that they're bad 😄 

https://github.com/digidem/mapeo-map-server/tree/master/test/fixtures/bad-mbtiles

achou11 committed:
fix ui freezing when import error occurs

achou11 committed:
use separate modal sheet for error states

achou11 commented:
I think i fixed the ui freezing issue with 38004d253168904657750350544c014f740f17eb

achou11 committed:
major refactor with gregor

achou11 commented:
as of 3ea59d2643978bcf23c2264cb8211c75d5420508, this is really a full-on refactor of the map server integration with mobile, to account for challenges presented by the import progress feature. don't have the energy to detail all of the changes but will try to do so tomorrow.

a couple of issues still remain, proper flows + updates when imports error or complete. there are some specific cases that aren't fully working yet

gmaclennan committed:
WIP: refactor & cleanup

gmaclennan committed:
WIP: BG maps refactor

gmaclennan committed:
WIP continued fixes

gmaclennan committed:
fix: update to RN v0.66.5

gmaclennan committed:
fix: update to react-native-v8@1.4.8

gmaclennan committed:
Merge branch 'fix/react-native-patch' into new-map-import-progress

* fix/react-native-patch:
  fix: update to react-native-v8@1.4.8
  fix: update to RN v0.66.5

gmaclennan commented:
fix(server): don't add event listeners on start

The map-server::state event listeners are added on server initialization,
so they should not be added and removed on start and stop.
```diff
diff --git a/src/backend/server.js b/src/backend/server.js
index 2ef1f87d..6644d5e3 100644
--- a/src/backend/server.js
+++ b/src/backend/server.js
@@ -267,9 +267,6 @@ function createServer({
           return;
         }
 
-        handleGetState = createGetStateHandler(mapServer);
-        rnBridge.channel.on("map-server::get-state", handleGetState);
-
         mapServer
           .start()
           .then(() => {
@@ -283,20 +280,6 @@ function createServer({
       stop: () => {
         log("Stopping map server");
 
-        if (handleGetState) {
-          rnBridge.channel.removeListener(
-            "map-server::get-state",
-            handleGetState
-          );
-        }
-
-        if (handleReportErrorState) {
-          rnBridge.channel.removeListener(
-            "map-server::get-state",
-            handleReportErrorState
-          );
-        }
-
         // TODO: Handle this better
         if (!mapServer) {
           log("Map server instance does not exist");

```

achou11 commented:
> fix(server): don't add event listeners on start
> 
> The map-server::state event listeners are added on server initialization, so they should not be added and removed on start and stop.
> 
> ```diff
> diff --git a/src/backend/server.js b/src/backend/server.js
> index 2ef1f87d..6644d5e3 100644
> --- a/src/backend/server.js
> +++ b/src/backend/server.js
> @@ -267,9 +267,6 @@ function createServer({
>            return;
>          }
>  
> -        handleGetState = createGetStateHandler(mapServer);
> -        rnBridge.channel.on("map-server::get-state", handleGetState);
> -
>          mapServer
>            .start()
>            .then(() => {
> @@ -283,20 +280,6 @@ function createServer({
>        stop: () => {
>          log("Stopping map server");
>  
> -        if (handleGetState) {
> -          rnBridge.channel.removeListener(
> -            "map-server::get-state",
> -            handleGetState
> -          );
> -        }
> -
> -        if (handleReportErrorState) {
> -          rnBridge.channel.removeListener(
> -            "map-server::get-state",
> -            handleReportErrorState
> -          );
> -        }
> -
>          // TODO: Handle this better
>          if (!mapServer) {
>            log("Map server instance does not exist");
> ```

little bit confused by this comment. the suggested removals are for `::get-state`, not `::state`. Is adding event listeners for that unnecessary?

gmaclennan committed:
fix: Got boolean wrong

gmaclennan committed:
fix(server): dont add/remove onState on start/stop

The map-server::state handler is added on initialization, so should not then
be added and removed on start and stop

gmaclennan commented:
> little bit confused by this comment. the suggested removals are for `::get-state`, not `::state`. Is adding event listeners for that unnecessary?

Sorry I meant to say `map-server::get-state`.

[Line #215](https://github.com/digidem/mapeo-mobile/blob/98de3b2c3f2ea0ed22849b827ec8ed6fb509b233/src/backend/server.js#L215) adds a `get-state` event listener during the `createMapServer` initialization, and [Line #233](https://github.com/digidem/mapeo-mobile/blob/98de3b2c3f2ea0ed22849b827ec8ed6fb509b233/src/backend/server.js#L233) adds the special error handler incase the map server did not initialize.

The code I deleted was:

A) During server start it was adding a second `get-state` listener, doing the same thing as the listener added during initialization.
B) During server stop it was removing the `get-state` listener, but since it was attached during initialization we don't need to add and remove it during start and stop.
C) The removal of the special error handler listener during stop was redundant because that code would never be reached (it was only attached if `mapServer === undefined`) but even if it was reached then we would not want this listener to be removed (imagine the user opens the app, map-server fails to start, so error listener attached, then user switches away (server stop) then switches back (server start), now the error listener is removed, so sending a `get-state` event from the client would not work as expected.

achou11 committed:
Merge branch 'develop' into new-map-import-progress

achou11 committed:
move fetch polyfills to polyfills directory

achou11 committed:
remove console log

achou11 committed:
Merge branch 'develop' into new-map-import-progress

achou11 commented on src/frontend/screens/Settings/MapSettings/MapImportBottomSheet.tsx:
```diff
@@ -0,0 +1,326 @@
+  );
+};
+
+const MapImportBottomSheet = React.forwardRef<MapImportBottomSheetMethods, {}>(
```
depending on the behavior we want, this could reuse our BottomSheetModal component. using the BottomSheet was to account for issues with collapsing/closing and re-opening the sheet in sequence, which was only necessary because of my initial implementation of the error-handling.

think with Gregor's refactor and the UX changes for errors on this screen, we could scrap the usage of the BottomSheet and stick with BottomSheetModal, since that more closely reflects the usage we want (sheet that has blocking actions)

achou11 commented on src/frontend/screens/Settings/MapSettings/MapImportBottomSheet.tsx:
```diff
@@ -0,0 +1,326 @@
+    const { snapPoints, updateSheetHeight } = useSnapPointsCalculator();
+    const { add: addMapImport } = useMapImportsManager();
+
+    React.useImperativeHandle(ref, () => {
```
wow first time i've seen useImperativeHandle being used. that's pretty cool! maybe `BottomSheetModal` should be doing something similar (not in this PR's scope, obviously)

achou11 committed:
remove unused import

achou11 committed:
fix frontend tests

achou11 committed:
Merge branch 'develop' into new-map-import-progress

achou11 committed:
Merge branch 'develop' into new-map-import-progress

achou11 committed:
elaborate on comment

achou11 committed:
style fixes for map info screen

achou11 committed:
fix background maps screen title

achou11 committed:
styles fixes for map selector bottom sheet on map screen

achou11 committed:
update @mapeo/map-server dep

achou11 committed:
remove unused import

achou11 commented on src/frontend/hooks/useMapStyles.ts:
```diff
@@ -0,0 +1,125 @@
+      return () => {
+        didCancel = true;
+      };
+    }, [activeImports, backgroundMaps])
```
i added this deps array in some previous commit because it would otherwise potentially infinitely render. does this look right @gmaclennan ? I think where this falls a little short is when mutative operations from the client occur (e.g. deleting a style), this doesn't really trigger. Could potentially define those actions in this hook and use some dummy state token to manually trigger this, I guess 🤔

achou11 commented on src/frontend/polyfills/fetch.ts:
```diff
@@ -0,0 +1,34 @@
+require("fast-text-encoding");
+
+// Polyfill ReadableStream
+const { ReadableStream } = require("web-streams-polyfill/ponyfill/es6");
+
+polyfillGlobal("ReadableStream", () => ReadableStream);
+
+const { fetch, Headers, Request, Response } = require("react-native-fetch-api");
+
+// Polyfill fetch
+polyfillGlobal("fetch", () => fetch);
+polyfillGlobal("Headers", () => Headers);
+polyfillGlobal("Request", () => Request);
+polyfillGlobal("Response", () => Response);
+
+// Polyfill window object (needed for @microsoft/fetch-event-source)
+polyfillGlobal("window", () => ({
+  ...(global.window || {}),
+  setTimeout,
+  clearTimeout,
+  fetch,
+}));
+
+// Polyfill document object (needed for @microsoft/fetch-event-source)
+polyfillGlobal("document", () => ({
+  ...(global.document || {}),
+  hidden: false,
+  addEventListener: () => {},
+  removeEventListener: () => {},
+}));
```
TODO: probably should write type defs that extend the global

achou11 commented:
Broke this PR up into 3 stacked ones. Hopefully makes it easier to review. Will keep this one open for now until someone says we should go with the stacked ones or this one 😄 

1. #1048
2. #1049
3. #1050

achou11 committed:
remove unused text-encoding dep

achou11 committed:
Merge branch 'develop' into new-map-import-progress

achou11 commented:
closing this in favor of the partitioned ones mentioned in https://github.com/digidem/mapeo-mobile/pull/1031#issuecomment-1317395713
